### PR TITLE
Improved exercise updates

### PR DESCRIPTION
**Pull request by nikke234**
# Description of the PR

This PR moves exercises updates and authentication out of `MainViewModel`. Along with the code improvements it also fixes two user-facing issues:

- Refreshing assignments no longer causes the UI to freeze (the update no longer happens in the UI thread)
- Module and assignment refreshes actually do different things.

The exercise update still has the previous approach of replacing the entire `List<ExerciseGroup>` with a new list, instead of reusing the existing objects. It's possible to instead just update the existing model instances (e.g. update the points of exercises, add any missing submissions, etc.), although I'm not sure how that would interact with filtering.

TODO: Unit tests

# Testing
<table>
  <tr>
    <th>unit</th>
    <th>integration</th>
    <th>e2e</th>
    <th>manual</th>
  </tr>
  <tr>
    <td>
      <ul>
        <li>- [ ] new <b>unit</b> tests created</li>
        <li>- [x] all <b>unit</b> tests pass</li>
      </ul>
    </td>
    <td>
      <ul>
        <li>- [ ] new <b>integration</b> tests created</li>
        <li>- [x] all <b>integration</b> tests pass</li>
      </ul>
    </td>
    <td>
      <ul>
        <li>- [ ] new <b>e2e</b> tests created</li>
        <li>- [x] all <b>e2e</b> tests pass</li>
      </ul>
    </td>
    <td>
      <ul>
        <li>- [ ] <b>manual</b> testing went well</li>
      </ul>
    </td>
  </tr>
</table>  
  
# Have you updated the [TESTING.md](https://github.com/Aalto-LeTech/intellij-plugin/blob/master/TESTING.md) or other relevant documentation on _your_ branch?

- [ ] Yes.
- [ ] Not yet. I will do it next.
- [x] Not relevant.

### CODE REVIEW

File: src/main/java/fi/aalto/cs/apluscourses/presentation/MainViewModel.java [hunk 1]
```diff
@@ -57,77 +57,24 @@
...
-   *
-   * <p>This method does anything only when it's called the first time for an instance. All the
-   * subsequent calls do nothing.</p>
-   *
-   * @param passwordStorage Password storage.
-   * @param factory         Authentication factory.
-   */
-  public void readAuthenticationFromStorage(@Nullable PasswordStorage passwordStorage,
-                                            @NotNull TokenAuthentication.Factory factory) {
-    if (hasTriedToReadAuthenticationFromStorage.getAndSet(true) || authentication.get() != null) {
```
Comment (jaakkonarhi) [inline src/main/java/fi/aalto/cs/apluscourses/presentation/MainViewModel.java:122]: Is the hasTriedToReadAuthenticationFromStorage in MainViewModel useless now?

Comment (nikke234) [reply src/main/java/fi/aalto/cs/apluscourses/presentation/MainViewModel.java:122]: Oops, yes, good catch. I wanted to just cut+paste all authentication stuff from MainViewModel -> CourseProject.

File: src/main/java/fi/aalto/cs/apluscourses/intellij/model/CourseUpdater.java [hunk 2]
```diff
@@ -20,14 +21,15 @@
 import java.util.HashSet;
```
Comment (jaakkonarhi) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/model/CourseUpdater.java:21]: unused import

File: src/main/java/fi/aalto/cs/apluscourses/intellij/model/CourseProject.java [hunk 3]
```diff
@@ -1,10 +1,20 @@
...
 
 import com.intellij.openapi.project.Project;
+import fi.aalto.cs.apluscourses.dal.PasswordStorage;
+import fi.aalto.cs.apluscourses.dal.TokenAuthentication;
+import fi.aalto.cs.apluscourses.model.Authentication;
 import fi.aalto.cs.apluscourses.model.Course;
+import fi.aalto.cs.apluscourses.model.ExerciseGroup;
 import fi.aalto.cs.apluscourses.utils.Event;
 import java.net.URL;
+import java.util.ArrayList;
```
Comment (jaakkonarhi) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/model/CourseProject.java:11]: unused import

**Review by jaakkonarhi — CHANGES_REQUESTED** [review]

File: src/main/java/fi/aalto/cs/apluscourses/intellij/model/ExercisesUpdater.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    var authentication = courseProject.getAuthentication();
+    if (authentication == null) {
+      return;
+    }
+    try {
+      var points = dataSource.getPoints(course, authentication);
+      if (Thread.interrupted()) {
+        return;
+      }
+      points.setSubmittableExercises(course.getExerciseModules().keySet()); // TODO: remove
```
Comment (Taikelenn) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/model/ExercisesUpdater.java:51]: Should this be removed if it's marked as "TODO: remove"? ;p

**Comment by Taikelenn** [comment]
While we're at changes regarding exercise trees, can we please make "Assignments" here lowercase:
無題

**Comment by Taikelenn** [comment]
And one more comment: if a user has entered an invalid token, then "Set your A+ token to view your assignments" should be shown in the assignments tree - this is how it is on the master branch. However, in this branch it shows "All assignments have been filtered out". Perhaps it's caused by `PluginSettings::registerCourseProject` and a call to `setAuthenticated` therein, namely an invalid authentication also causes the project to be considered "authenticated"

File: src/main/java/fi/aalto/cs/apluscourses/intellij/model/ExercisesUpdater.java (hunk 4 above)
Comment (nikke234) [reply src/main/java/fi/aalto/cs/apluscourses/intellij/model/ExercisesUpdater.java:51]: Yes eventually, but this is copied from `MainViewModel`, so the comment is still relevant but it's a bigger fix :smile:

**Comment by nikke234** [comment]
> And one more comment: if a user has entered an invalid token, then "Set your A+ token to view your assignments" should be shown in the assignments tree - this is how it is on the master branch. However, in this branch it shows "All assignments have been filtered out". Perhaps it's caused by `PluginSettings::registerCourseProject` and a call to `setAuthenticated` therein, namely an invalid authentication also causes the project to be considered "authenticated" (that's only my speculation)

Yes good finding, I will fix this

**Review by jaakkonarhi — APPROVED** [review]
Your code is perfect

**Review by Taikelenn — APPROVED** [review]
Acceptable

**Pull request merged**